packages/api/src/command/medical/cq-directory/create-cq-directory-entry.ts

```diff
@@ -1,32 +1,114 @@
+  dialect: dbCreds.engine,
+});
+
+export type CQOrganizationCreateResponse = { org: CQDirectoryEntry } & { updated: boolean };
+export type CQOrganizatioBulkCreateResponse = { updated: number; added: number };
+export type CQDirectoryEntryDataWithUpdateAndId = CQDirectoryEntryData & {
+  update?: boolean;
+  new?: boolean;
 };
 
 export const createOrUpdateCQDirectoryEntry = async (
   orgData: CQDirectoryEntryData
 ): Promise<CQOrganizationCreateResponse> => {
-  // ensure we never create more than one entry per cq org
   const existingOrg = await getCQDirectoryEntry({ oid: orgData.oid });
   if (existingOrg) {
     const updOrg = await updateCQDirectoryEntry({ ...orgData, id: existingOrg.id });
     return { org: updOrg, updated: true };
   }
 
-  const org = await createDirectoryDirectoryEntry(orgData);
+  const org = await createCQDirectoryEntry(orgData);
   return { org, updated: false };
 };
 
-async function createDirectoryDirectoryEntry(
-  orgData: CQDirectoryEntryData
-): Promise<CQDirectoryEntry> {
-  return await CQDirectoryEntryModel.create({
-    id: uuidv7(),
-    ...orgData,
+const createCQDirectoryEntry = async (orgData: CQDirectoryEntryData): Promise<CQDirectoryEntry> => {
+  return await CQDirectoryEntryModel.create(orgData);
+};
+
+export const createOrUpdateCQDirectoryEntries = async (
+  orgDataArray: CQDirectoryEntryDataWithUpdateAndId[]
+): Promise<CQOrganizatioBulkCreateResponse> => {
+  const oids = orgDataArray.map(data => data.oid);
+  const existingEntries = await getCQDirectoryEntriesIdsAndLastUpdated(oids); // TODO: could parallelize this
+
+  orgDataArray.forEach(orgData => {
+    const existingEntry = existingEntries.find(entry => entry.oid === orgData.oid);
+    if (existingEntry) {
+      const latestVersion = existingEntry.lastUpdated === orgData.lastUpdated;
+      if (!latestVersion) {
+        orgData.id = existingEntry.id;
+        orgData.update = true;
+      }
+    } else {
+      orgData.new = true;
+    }
   });
+
+  const newEntries = orgDataArray.filter(org => org.new);
+  const numNewEntries = newEntries.length;
+
+  const updEntries = orgDataArray.filter(org => org.update);
+  const numUpdEntries = updEntries.length;
+
+  if (numNewEntries) await createCQDirectoryEntries(newEntries);
+  if (numUpdEntries) await updateCQDirectoryEntries(updEntries);
+
+  return { added: numNewEntries, updated: numUpdEntries };
+};
+
+async function computeEarthPoint(orgData: CQDirectoryEntryData): Promise<string | undefined> {
+  if (orgData.lat && orgData.lon) {
+    const query = "SELECT ll_to_earth(:lat, :lon) as point";
+    const pointQueryResult = await sequelize.query(query, {
+      replacements: { lat: orgData.lat, lon: orgData.lon },
+      type: QueryTypes.SELECT,
+      logging: false,
+    });
+    const point = QueryResultSchema.parse(pointQueryResult);
+    return point[0].point;
+  }
+  return;
 }
+
+const createCQDirectoryEntries = async (orgDataArray: CQDirectoryEntryData[]): Promise<void> => {
```

Comment:
use promise.all to parallelize

Reply:
If we decide to parallelize - which is a good point, just consider we need to limit the amount of DB connections we use and the amount of load we incur to the DB. `executeAsynchronously` could help there.